doc/aggregation.md

```diff
@@ -0,0 +1,164 @@
+- No `xFilesFactor` entity: each point will be aggregated
+
+### Fetching data: before September 2019 (current `internal-aggregation = false` behavior)
+
+```sql
+SELECT Path, Time, Value, Timestamp
+FROM data WHERE ...
+```
+
+- Select all points
+- Aggregate them in *graphite-clickhouse* to the proper `archive` step
+- Pass further to *graphite-web*/*carbonapi*
+
+Problems:
+
+- A big overhead for Path (the heaviest part)
+- Overkill by network traffic, especially when CH cluster is used
+  - The CH node `query-initiator` must collect the whole data (in memory or on the disk), and only then the date will be passed further
+
+### Fetching data: after September 2019 ([#61](https://github.com/lomik/graphite-clickhouse/pull/61), [#62](https://github.com/lomik/graphite-clickhouse/pull/62), [#65](https://github.com/lomik/graphite-clickhouse/pull/61))
+
+```sql
+SELECT Path,
+  groupArray(Time),
+  groupArray(Value),
+  groupArray(Timestamp)
+FROM data WHERE ... GROUP BY Path
+```
+
+- Network consumption was decreased up to 6 times
+- But still selects all points and aggregates in *graphite-clickhouse*
+
+### Fetching data: September 2020 ([#88](https://github.com/lomik/graphite-clickhouse/pull/88))
+
+```sql
+SELECT Path,
+  arrayFilter(x->isNotNull(x),
+    anyOrNullResample($from, $until, $step)
+      (toUInt32(intDiv(Time, $step)*$step), Time)
+  ),
+  arrayFilter(x->isNotNull(x),
+    ${func}OrNullResample($from, $until, $step)
+      (Value, Time)
+  )
+FROM data WHERE ... GROUP BY Path
+```
+
+- This solution implements `archive` analog on CH side
+- The most of data is aggregated on CH shards and doesn't leave them, so `query-initiator` consumes much less memory
+- Together with *carbonapi* the `/render?maxDatePoints=x` parameter as well processed on CH side
+
+### Fetching data: April 2021 ([#145](https://github.com/lomik/graphite-clickhouse/pull/145))
+The query is modified a little bit, it doesn't use -OrNull anymore and compatible with more CH versions
+
+```sql
+WITH anyResample($from, $until, $step)(toUInt32(intDiv(Time, $step)*$step), Time) AS mask
+SELECT Path,
+ arrayFilter(m->m!=0, mask) AS times,
+ arrayFilter((v,m)->m!=0, ${func}Resample($from, $until, $step)(Value, Time), mask) AS values
+FROM data WHERE ... GROUP BY Path
+```
+
+## Fetching data: concepts' difference
+
+For small requests, the difference is not so big, but for the heavy one the amount of data was decreased up to 100 times:
+
+```
+target=${986_metrics_60s_precision}
+from=-7d
+maxDataPoints=100
+```
+
+| method     | rows    | points  | data (binary)    | time (s) |
+| -          | -       | -       | -                | -        |
+| row/point  | 9887027 | 9887027 | 556378258 (530M) | 16.486   |
+| groupArray | 986     | 9887027 | 158180388 (150M) | 35.498   |
+| -Resample  | 986     | 98553   | 1421418 (1M)     | 13.181   |
+
+*note*: it's localhost, so network data transfer overhead is not taken into account.
```

Comment:
`CH runs on the same server, so network transfer overhead is negligible`

Reply:
not actually what I've tried to say. But I'll try to rephrase it